File: packages/contracts/src/dsh.ts
import { canonical, deny, rawDigest, validateWithProfile } from './index.js';
import { safePath } from './release.js';
type Obj = Record<string, any>;
/** The digest and lock identity come from reviewed local bootstrap policy, never the candidate. */
export class TrustedDshClosure {
  private constructor(readonly identity:string, readonly lock:string, private entries:Map<string,Obj>) {}
  static verify(bytes:Uint8Array, expectedDigest:string, expectedLock:string):TrustedDshClosure {
    if(!/^[a-f0-9]{64}$/.test(expectedDigest) || !/^[a-f0-9]{64}$/.test(expectedLock) || rawDigest(bytes)!==expectedDigest) deny('DSH_TRUST');
    const closure=validateWithProfile('maestro.dsh.closure/1',bytes,true,'dsh-release') as Obj;
    if(closure.lock_sha256!==expectedLock) deny('DSH_LOCK');
    const entries=new Map<string,Obj>(), folded=new Set<string>();
    for(const entry of closure.files) {
      safePath(entry.path,true);
      if(folded.has(entry.path.toLowerCase()) || entry.role!=='dsh') deny('DSH_CLOSURE');
      folded.add(entry.path.toLowerCase());entries.set(entry.path,entry);
    }
    return new TrustedDshClosure(expectedDigest,expectedLock,entries);
  }
  check(manifest:Obj):void {
    if(manifest.dsh_closure!==this.identity || manifest.dsh_lock.sha256!==this.lock) deny('DSH_BINDING');
    const owned=new Map<string,string>(), inventory=new Map<string,Obj>();
    for(const entry of manifest.files) {
      if(inventory.has(entry.path)) deny('INVENTORY');
      inventory.set(entry.path,entry);
    }
    for(const component of manifest.components) for(const path of component.files) {
      if(owned.has(path) || !inventory.has(path)) deny('OWNERSHIP');
      owned.set(path,component.name);
    }
    if(owned.size!==inventory.size || inventory.size>32768 || inventory.size-this.entries.size>4096) deny('INVENTORY');
    for(const [path,entry] of this.entries) {
      if(!inventory.has(path) || !canonical(inventory.get(path)).equals(canonical(entry)) || owned.get(path)!=='deepseek-harness') deny('DSH_CLOSURE');
    }
    for(const [path,owner] of owned) if(owner==='deepseek-harness' && !this.entries.has(path)) deny('DSH_CLOSURE');
  }
}

File: crates/maestro-release/src/dsh.rs
use crate::{field, staged::safe_path_profile};
use maestro_contracts::{AdmissionProfile, Error, Result, raw_digest, validate_with_profile};
use serde_json::Value;
use std::collections::{BTreeMap, HashSet};

/// Expected identities are independent reviewed bootstrap inputs, not manifest claims.
pub struct TrustedDshClosure {
    identity: String,
    lock: String,
    entries: BTreeMap<String, Value>,
}
impl TrustedDshClosure {
    pub fn verify(bytes: &[u8], expected_digest: &str, expected_lock: &str) -> Result<Self> {
        let hash = |v: &str| {
            v.len() == 64
                && v.bytes()
                    .all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b))
        };
        if !hash(expected_digest) || !hash(expected_lock) || raw_digest(bytes) != expected_digest {
            return Err(Error("DSH_TRUST"));
        }
        let closure = validate_with_profile(
            "maestro.dsh.closure/1",
            bytes,
            true,
            AdmissionProfile::DshRelease,
        )?;
        if field(&closure, "lock_sha256")? != expected_lock {
            return Err(Error("DSH_LOCK"));
        }
        let mut entries = BTreeMap::new();
        let mut folded = HashSet::new();
        for entry in closure["files"].as_array().ok_or(Error("DSH_CLOSURE"))? {
            let path = field(entry, "path")?;
            safe_path_profile(path, true)?;
            if !folded.insert(path.to_ascii_lowercase()) || entry["role"] != "dsh" {
                return Err(Error("DSH_CLOSURE"));
            }
            entries.insert(path.to_owned(), entry.clone());
        }
        Ok(Self {
            identity: expected_digest.to_owned(),
            lock: expected_lock.to_owned(),
            entries,
        })
    }
    pub fn check(&self, manifest: &Value) -> Result<()> {
        if manifest["dsh_closure"] != self.identity || manifest["dsh_lock"]["sha256"] != self.lock {
            return Err(Error("DSH_BINDING"));
        }
        let mut inventory = BTreeMap::new();
        let mut owned = BTreeMap::new();
        for entry in manifest["files"].as_array().ok_or(Error("INVENTORY"))? {
            if inventory.insert(field(entry, "path")?, entry).is_some() {
                return Err(Error("INVENTORY"));
            }
        }
        for component in manifest["components"]
            .as_array()
            .ok_or(Error("OWNERSHIP"))?
        {
            for path in component["files"].as_array().ok_or(Error("OWNERSHIP"))? {
                let path = path.as_str().ok_or(Error("OWNERSHIP"))?;
                if owned.insert(path, field(component, "name")?).is_some()
                    || !inventory.contains_key(path)
                {
                    return Err(Error("OWNERSHIP"));
                }
            }
        }
        if owned.len() != inventory.len()
            || inventory.len() > 32768
            || inventory.len().saturating_sub(self.entries.len()) > 4096
        {
            return Err(Error("INVENTORY"));
        }
        for (path, entry) in &self.entries {
            if inventory.get(path.as_str()).copied() != Some(entry)
                || owned.get(path.as_str()).copied() != Some("deepseek-harness")
            {
                return Err(Error("DSH_CLOSURE"));
            }
        }
        for (path, owner) in owned {
            if owner == "deepseek-harness" && !self.entries.contains_key(path) {
                return Err(Error("DSH_CLOSURE"));
            }
        }
        Ok(())
    }
}
